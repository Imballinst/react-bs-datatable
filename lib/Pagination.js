"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ButtonGroup = _interopRequireDefault(require("react-bootstrap/ButtonGroup"));

var _Button = _interopRequireDefault(require("react-bootstrap/Button"));

var _NavButton = _interopRequireDefault(require("./modules/NavButton"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Pagination(_ref) {
  var data = _ref.data,
      rowsPerPage = _ref.rowsPerPage,
      keyName = _ref.keyName,
      currentPage = _ref.currentPage,
      onPageNavigate = _ref.onPageNavigate,
      labels = _ref.labels;
  var renderedElements = null;

  if (rowsPerPage !== undefined) {
    var dataLength = data.length;
    var numberOfPages = Math.ceil(dataLength / rowsPerPage);
    var buttons = [];
    var firstLabel = labels.first || 'First';
    var prevLabel = labels.prev || 'Prev';
    var nextLabel = labels.next || 'Next';
    var lastLabel = labels.last || 'Last';
    var startNumber;
    var i = 0;
    var hasPrev = true;
    var hasNext = true;

    if (currentPage === 1) {
      startNumber = 1;
      hasPrev = false;
      hasNext = numberOfPages > 1;
    } else if (currentPage === numberOfPages && numberOfPages !== 1) {
      startNumber = numberOfPages - 2 > 0 ? currentPage - 2 : 1;
      hasNext = false;
    } else {
      startNumber = currentPage - 1;
    }

    buttons.push(_react.default.createElement(_NavButton.default, {
      key: "".concat(keyName, "-page-").concat(firstLabel),
      pageNavNumber: 1,
      disabled: !hasPrev,
      onPageNavigate: onPageNavigate,
      label: firstLabel
    }), _react.default.createElement(_NavButton.default, {
      key: "".concat(keyName, "-page-").concat(prevLabel),
      pageNavNumber: currentPage - 1,
      disabled: !hasPrev,
      onPageNavigate: onPageNavigate,
      label: prevLabel
    }));

    while (i < 3 && startNumber <= numberOfPages) {
      var pageBtnProps = {
        key: "".concat(keyName, "-page-btn-").concat(startNumber),
        onClick: onPageNavigate(startNumber),
        active: currentPage === startNumber
      };
      buttons.push(_react.default.createElement(_Button.default, pageBtnProps, startNumber));
      i += 1;
      startNumber += 1;
    }

    buttons.push(_react.default.createElement(_NavButton.default, {
      key: "".concat(keyName, "-page-").concat(nextLabel),
      pageNavNumber: currentPage + 1,
      disabled: !hasNext,
      onPageNavigate: onPageNavigate,
      label: nextLabel
    }), _react.default.createElement(_NavButton.default, {
      key: "".concat(keyName, "-page-").concat(lastLabel),
      pageNavNumber: numberOfPages,
      disabled: !hasNext,
      onPageNavigate: onPageNavigate,
      label: lastLabel
    }));
    renderedElements = _react.default.createElement(_ButtonGroup.default, {
      className: "btn-group-page-nav"
    }, buttons);
  }

  return renderedElements;
}

Pagination.propTypes = {
  data: _propTypes.default.array.isRequired,
  keyName: _propTypes.default.string.isRequired,
  currentPage: _propTypes.default.number.isRequired,
  onPageNavigate: _propTypes.default.func.isRequired,
  rowsPerPage: _propTypes.default.number,
  labels: _propTypes.default.object
};
Pagination.defaultProps = {
  rowsPerPage: undefined,
  labels: {}
};
var _default = Pagination;
exports.default = _default;