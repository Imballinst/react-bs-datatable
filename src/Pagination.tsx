import React, { useCallback } from 'react';

import ButtonGroup from 'react-bootstrap/ButtonGroup';

import NavButton from './modules/NavButton';
import { LabelType, RowsPerPageType, TableClasses } from './helpers/types';

type PaginationProps = {
  rowsPerPage: RowsPerPageType;
  currentPage: number;
  maxPage?: number;
  onPageNavigate: any;
  labels: LabelType;
  classes: TableClasses;
};

export default function Pagination({
  rowsPerPage,
  currentPage,
  onPageNavigate,
  labels,
  maxPage,
  classes
}: PaginationProps) {
  const paginateHandler = useCallback(
    (pageNum: number) => {
      return () => onPageNavigate(pageNum);
    },
    [onPageNavigate]
  );

  let renderedElements = null;

  if (rowsPerPage !== undefined && maxPage !== undefined) {
    const buttons = [];

    const firstLabel = labels.first || 'First';
    const prevLabel = labels.prev || 'Prev';
    const nextLabel = labels.next || 'Next';
    const lastLabel = labels.last || 'Last';

    const isCurrentPageOutOfBounds = currentPage > maxPage;
    let startNumber;
    let i = 0;
    let hasPrev = true;
    let hasNext = true;

    if (currentPage === 1) {
      // Active button is the first one.
      startNumber = 1;
      hasPrev = false;
      hasNext = maxPage > 1;
    } else if (currentPage === maxPage && maxPage !== 1) {
      // Active button is in the last.
      startNumber = maxPage - 2 > 0 ? currentPage - 2 : 1;
      hasNext = false;
    } else {
      // Active button is in the middle.
      startNumber = currentPage - 1;
    }

    buttons.push(
      <NavButton
        className={classes.paginationButton}
        key={`page-${firstLabel}`}
        pageNumber={1}
        disabled={!hasPrev}
        onPageNavigate={paginateHandler}
        label={firstLabel}
      />,
      <NavButton
        className={classes.paginationButton}
        key={`page-${prevLabel}`}
        // If out of bounds, prev button refers to the last page.
        pageNumber={isCurrentPageOutOfBounds ? maxPage : currentPage - 1}
        disabled={!hasPrev}
        onPageNavigate={paginateHandler}
        label={prevLabel}
      />
    );

    if (!isCurrentPageOutOfBounds) {
      while (i < 3 && startNumber <= maxPage) {
        const pageBtnProps = {
          key: `page-btn-${i}`,
          pageNumber: startNumber,
          disabled: currentPage === startNumber,
          onPageNavigate: paginateHandler,
          label: startNumber,
          className: classes.paginationButton
        };

        buttons.push(<NavButton {...pageBtnProps} />);

        i += 1;
        startNumber += 1;
      }
    } else {
      // If current page is more than maxPage, we disable the button.
      // This is intentional so the user knows that no data in there.
      hasNext = false;
      buttons.push(
        <NavButton
          key={`page-btn-${startNumber}`}
          className={classes.paginationButton}
          pageNumber={currentPage - 1}
          disabled
          onPageNavigate={paginateHandler}
          label={currentPage}
        />
      );
    }

    buttons.push(
      <NavButton
        className={classes.paginationButton}
        key={`page-${nextLabel}`}
        pageNumber={currentPage + 1}
        disabled={!hasNext}
        onPageNavigate={paginateHandler}
        label={nextLabel}
      />,
      <NavButton
        className={classes.paginationButton}
        key={`page-${lastLabel}`}
        pageNumber={maxPage}
        disabled={!hasNext}
        onPageNavigate={paginateHandler}
        label={lastLabel}
      />
    );

    renderedElements = (
      <ButtonGroup className={classes.paginationButtonGroup}>
        {buttons}
      </ButtonGroup>
    );
  }

  return renderedElements;
}
