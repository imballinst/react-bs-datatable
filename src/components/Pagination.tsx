import React, { useEffect, useMemo, useRef } from 'react';
import { Button, ButtonGroup } from 'react-bootstrap';
import { useControlledStateSetter } from '../helpers/hooks';

import { makeClasses } from '../helpers/object';
import { PaginationOnChange } from '../helpers/types';
import { useDatatableWrapper } from './DatatableWrapper';
import { getPageNumbers } from '../helpers/pagination';

const DEFAULT_PAGINATION_RANGE = 3;

/** This is an interface to customize the pagination labels. */
export interface PaginationLabels {
  /** The "First" button label. Defaults to "First". */
  firstPage?: string;
  /** The "Last" button label. Defaults to "Last". */
  lastPage?: string;
  /** The "Prev" button label. Defaults to "Prev". */
  prevPage?: string;
  /** The "Next" button label. Defaults to "Next". */
  nextPage?: string;
}

/**
 * This is an interface for customizing the classes for
 * the `Pagination` component.
 */
export interface PaginationClasses {
  /** The class for each of the pagination button. */
  button?: string;
  /** The class for the pagination button group. */
  buttonGroup?: string;
}

/**
 * This is an interface for `Pagination` component props.
 */
export interface PaginationProps {
  /** Customize the labels of the `Pagination` component. */
  labels?: PaginationLabels;
  /** Customize the classes of the `Pagination` component. */
  classes?: PaginationClasses;
  /**
   * Determine whether the pagination button group should be always visible or not.
   * When set to `false`, the pagination will be hidden when there is only 1 page.
   * To prevent layout shifts, `visibility: hidden` will be applied instead of
   * `display: none` style. Defaults to `true`.
   */
  alwaysShowPagination?: boolean;
  /**
   * The pagination range that shows between the "First"/"Prev" and "Next"/"Last" buttons.
   * Defaults to 3, for example: First, Prev, 1, 2, 3, Next, Last.
   * When set to 5, for example: First, Prev, 1, 2, 3, 4, 5 Next, Last.
   **/
  paginationRange?: number;
  /** Props to make the component controlled. */
  controlledProps?: {
    /** The currently active page. */
    currentPage?: number;
    /**
     * This is used to determine * the last numbered button in the
     * pagination button group. This also determines the next page number
     * when the "Last" button is clicked.
     */
    maxPage?: number;
    /** The function fired when any of the pagination buttons is clicked. */
    onPaginationChange?: PaginationOnChange;
  };
}

/**
 * Renders a control for the pagination. It consists of "First", "Prev",
 * "Next", and "Last" buttons. Additionally, between the "Prev" and "Next" buttons,
 * there will be at maximum 3 numbered buttons, indicating the pages to navigate.
 *
 * The currently active page will always be disabled, as well as the "First" or "Last"
 * buttons if the `currentPage` represents the first page and last page, respectively.
 *
 * When `alwaysShowPagination` is set to `false`, then this component will be visually hidden.
 */
export function Pagination({
  labels,
  classes,
  paginationRange = DEFAULT_PAGINATION_RANGE,
  alwaysShowPagination = true,
  controlledProps
}: PaginationProps) {
  const {
    currentPageState: currentPageStateContext,
    maxPage: maxPageContext,
    onPaginationChange: onPaginationChangeContext
  } = useDatatableWrapper();
  useControlledStateSetter(controlledProps);

  // Controlled has the bigger priority.
  const currentPageState =
    controlledProps?.currentPage || currentPageStateContext;
  const maxPage = controlledProps?.maxPage || maxPageContext;
  const onPaginationChange =
    controlledProps?.onPaginationChange || onPaginationChangeContext;

  // Hide pagination if pagination is meant to not always be shown and the max page is 1,
  // or if rows per page is -1.
  const hidePagination = !alwaysShowPagination && maxPage === 1;

  const buttons = [];
  const firstLabel = labels?.firstPage || 'First';
  const prevLabel = labels?.prevPage || 'Prev';
  const nextLabel = labels?.nextPage || 'Next';
  const lastLabel = labels?.lastPage || 'Last';

  const prevPageNumbers = useRef<number[] | undefined>(undefined);
  const pageNumbers = useMemo(() => {
    return getPageNumbers({
      currentPageState,
      maxPage,
      paginationRange,
      prevPageNumbers: prevPageNumbers.current
    });
  }, [currentPageState, maxPage, paginationRange]);

  useEffect(() => {
    prevPageNumbers.current = pageNumbers;
  }, [pageNumbers]);

  const hasPrev = currentPageState !== 1;
  const hasNext = currentPageState !== maxPage;

  buttons.push(
    <Button
      className={classes?.button}
      key={`page-${firstLabel}`}
      onClick={() => onPaginationChange(1)}
      aria-label="Go to first page"
      disabled={!hasPrev}
    >
      {firstLabel}
    </Button>,
    <Button
      className={classes?.button}
      key={`page-${prevLabel}`}
      aria-label="Go to previous page"
      // If out of bounds, prev button refers to the last page.
      onClick={() => onPaginationChange(currentPageState - 1)}
      disabled={!hasPrev}
    >
      {prevLabel}
    </Button>,
    ...pageNumbers.map((pageNumber) => (
      <Button
        key={`page-btn-${pageNumber}`}
        onClick={() => onPaginationChange(pageNumber)}
        disabled={currentPageState === pageNumber}
        aria-label={`Go to page ${pageNumber}`}
        className={classes?.button}
      >
        {pageNumber}
      </Button>
    )),
    <Button
      className={classes?.button}
      key={`page-${nextLabel}`}
      onClick={() => onPaginationChange(currentPageState + 1)}
      aria-label="Go to next page"
      disabled={!hasNext}
    >
      {nextLabel}
    </Button>,
    <Button
      className={classes?.button}
      key={`page-${lastLabel}`}
      onClick={() => onPaginationChange(maxPage)}
      aria-label="Go to last page"
      disabled={!hasNext}
    >
      {lastLabel}
    </Button>
  );

  return (
    <ButtonGroup
      aria-label="Pagination button group"
      className={makeClasses('ButtonGroup__root', classes?.buttonGroup, {
        invisible: hidePagination
      })}
    >
      {buttons}
    </ButtonGroup>
  );
}
