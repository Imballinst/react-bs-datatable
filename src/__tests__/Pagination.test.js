import React from 'react';
import { shallow } from 'enzyme';

import Pagination from '../Pagination';

const onPageNavigate = jest.fn();

function setup() {
  const props = {
    data: [],
    keyName: 'pagination-keyname',
    currentPage: 1,
    onPageNavigate,
    rowsPerPage: 5
  };

  const enzymeWrapper = shallow(<Pagination {...props} />);

  return {
    props,
    enzymeWrapper
  };
}

describe('Pagination component (src/Pagination)', () => {
  it('should match normal snapshot', () => {
    const { props, enzymeWrapper } = setup();

    expect(enzymeWrapper).toMatchSnapshot();
  });

  it('should have the same props before and after render', () => {
    const { props, enzymeWrapper } = setup();

    expect(enzymeWrapper.prop('data')).toEqual([]);
    expect(enzymeWrapper.prop('keyName')).toBe('pagination-keyname');
    expect(enzymeWrapper.prop('currentPage')).toBe(1);
    expect(enzymeWrapper.prop('rowsPerPage')).toBe(5);
    expect(enzymeWrapper.prop('onPageNavigate')).toBe(onPageNavigate);
  });
});
