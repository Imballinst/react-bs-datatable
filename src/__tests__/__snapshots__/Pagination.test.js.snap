// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component (src/Pagination) should match normal snapshot 1`] = `ShallowWrapper {}`;
